touchforms/backend/touchcare.py

```diff
@@ -31,263 +33,49 @@
-        self.fully_loaded = True
-
-    def load_object_ids(self):
-        if self.form_context.get('all_case_ids', None):
-            case_ids = self.form_context.get('all_case_ids')
-        else:
-            case_ids = query_case_ids(self.query_func, criteria=self.additional_filters)
-        # todo: see note above about why sorting is necessary
-        self.ids = dict(enumerate(sorted(case_ids)))
-
-    def getIDsForValue(self, field_name, value):
-        logger.debug('case index lookup %s %s' % (field_name, value))
-
-        if (field_name, value) not in self.cached_lookups:
-            if field_name == 'case-id':
-                cases = [self.read_object(value)]
-            else:
-                try:
-                    get_val = {
-                        'case-type': lambda c: c.getTypeId(),
-                        'case-status': lambda c: 'closed' if c.isClosed() else 'open',
-                    }[field_name]
-                except KeyError:
-                    # Try any unrecognized field name as a case id field.
-                    # Needed for 'case-in-goal' lookup in PACT Care Plan app.
-                    cases = [self.read_object(value)]
-                else:
-                    cases = [c for c in self.objects.values() if get_val(c) == value]
-
-            self.cached_lookups[(field_name, value)] = cases
-
-        cases = self.cached_lookups[(field_name, value)]
-        id_map = dict((v, k) for k, v in self.ids.iteritems())
-        try:
-            return to_vect(id_map[c.getCaseId()] for c in cases)
-        except KeyError:
-            # Case was not found in id_map
-            raise CaseNotFound
-
-
-class LedgerDatabase(TouchformsStorageUtility):
-    def __init__(self, host, domain, auth, additional_filters, preload):
-        super(LedgerDatabase, self).__init__(host, domain, auth, additional_filters, preload)
-
-    def get_object_id(self, ledger):
-        return ledger.getEntiyId()
-
-    def fetch_object(self, entity_id):
-        return query_ledger_for_case(self.query_func, entity_id)
-
-    def load_object_ids(self):
-        case_ids = query_case_ids(self.query_func, criteria=self.additional_filters)
-        self.ids = dict(enumerate(case_ids))
-
-    def getIDsForValue(self, field_name, value):
-        logger.debug('ledger lookup %s %s' % (field_name, value))
-        if (field_name, value) not in self.cached_lookups:
-            if field_name == 'entity-id':
-                ledgers = [self.read_object(value)]
-            else:
-                raise NotImplementedError("Only entity-id lookup is currently supported!")
-
-            self.cached_lookups[(field_name, value)] = ledgers
-
-        else:
-            ledgers = self.cached_lookups[(field_name, value)]
+            self.last_sync = self.sandbox.getLastSync()
+        except:
+            logger.error("Unable to get last sync for usertime for user %s " % self.username)
```

Comment:
That's a nice trick, cool

On Sunday, August 2, 2015, Simon Kelly <notifications@github.com> wrote:

> In touchforms/backend/touchcare.py
> <https://github.com/dimagi/touchforms/pull/142#discussion_r36043026>:
>
> > -
> > -    def getIDsForValue(self, field_name, value):
> > -        logger.debug('ledger lookup %s %s' % (field_name, value))
> > -        if (field_name, value) not in self.cached_lookups:
> > -            if field_name == 'entity-id':
> > -                ledgers = [self.read_object(value)]
> > -            else:
> > -                raise NotImplementedError("Only entity-id lookup is currently supported!")
> > -
> > -            self.cached_lookups[(field_name, value)] = ledgers
> > -
> > -        else:
> > -            ledgers = self.cached_lookups[(field_name, value)]
> > +            self.last_sync = self.sandbox.getLastSync()
> > +        except:
> > +            logger.error("Unable to get last sync for usertime for user %s " % self.username)
>
> if you use logger.exception then it will include the exception info for
> you (no need to put the error in the message.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/touchforms/pull/142/files#r36043026>.
>